pkg/pillar/utils/deserialize.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2017-2023 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package utils
+
+import (
+	"encoding/json"
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	"io"
+	"os"
+	"reflect"
+)
+
+func readFile(filename string) ([]byte, error) {
+	// Check if the file exists
+	if _, err := os.Stat(filename); os.IsNotExist(err) {
+		log.Errorf("File %s does not exist", filename)
+		return nil, err
+	}
+
+	// Open the file
+	file, err := os.Open(filename)
+	if err != nil {
+		log.Errorf("Failed to open file %s, %s", filename, err.Error())
+		return nil, err
+	}
+	defer file.Close()
+
+	// Read the data
+	data, err := io.ReadAll(file)
+	if err != nil {
+		log.Errorf("Failed to read data from file %s, %s", filename, err.Error())
+		return nil, err
+	}
+
+	return data, nil
+}
+
+func validateFields(dataMap map[string]interface{}, expectedFields map[string]bool, criticalFields map[string]bool, filename string) error {
+	for k := range dataMap {
+		if _, ok := expectedFields[k]; !ok {
+			log.Warnf("Unexpected field %s in stored data in %s", k, filename)
+		}
+	}
+
+	for k := range expectedFields {
+		if _, ok := dataMap[k]; !ok {
+			if criticalFields[k] {
+				errMsg := fmt.Sprintf("Critical field %s missing in stored data in %s", k, filename)
+				log.Errorf(errMsg)
+				return fmt.Errorf(errMsg)
+			}
+			log.Warnf("Missing field %s in stored data in %s", k, filename)
+		}
+	}
+
+	return nil
+}
+
+func extractFields(t reflect.Type, fieldsMap *map[string]bool, criticalFields *map[string]bool) {
```

Comment:
Maps are passed by reference, you do not need to pass them with pointer.

Reply:
Thanks! I will fix it!

Reply:
Done